Add chart for Traefik-based ingress controller

krancour opened the pull request:
Current Todo:

- [ ] Decide on whether this chart should be in stable or incubator

krancour committed:
Add chart for Traefik-based ingress controller

prydonius commented:
Thanks @krancour!

prydonius commented on incubator/traefik/README.md:
```diff
@@ -0,0 +1,82 @@
+## Introduction
+
+This chart bootstraps Traefik as a Kubernetes ingress controller with optional support for SSL and
+Let's Encrypt. By default, Traefik will be installed into the `kube-system` namespace.
```
Is this a requirement? I think we generally prefer that Helm manages the namespace Charts get installed into with the `--namespace` flag.

krancour replied:
It's not a hard and fast requirement, however, since ingress controllers are frequently used to watch `Ingress` objects _cluster-wide_ (i.e. in all namespaces), they often require the elevated permissions of the `kube-system` namespace's default service account.

When writing this chart, I assumed that for the less-frequent (or so I imagine) use case where someone _does_ want to install this into some other namespace that `--namespace` would _override_ the hard-coded default of `kube-system`. I've put that to the test now, however, and found it doesn't work that way!

cc @technosophos @michelleN isn't `--namespace` supposed to override whatever default is hard-coded in the chart's manifests? Or was it just wishful thinking on my part that it works that way?

At any rate, it does seem that what is best for now is that I omit this hard-coded default and instead _document_ that this component would most often be installed into `kube-system`.

krancour replied:
@prydonius I've amended the PR as discussed above.

krancour committed:
Do not hard-code the namespace in the traefik chart

chrislovecnm commented on incubator/traefik/LICENSE:
```diff
@@ -0,0 +1,21 @@
+MIT License
+
+Copyright (c) 2016 Deis
```
Do we need this? This project is Apache 2 licensed and now we have MIT?

krancour replied:
Well, that's a tricky question. It seems individual charts are subject to their own license, no? What I opted to do here was use the same license that Traefik itself uses. If that's the wrong move, lmk.

viglesiasce replied:
I think this code/repo can have a separate license from the source of traefik itself.

krancour replied:
Ok. I'll change it.

krancour replied:
I'm actually just going to remove the license from this chart altogether. It seems license is optional for charts and no other chart in this repo includes one. I'm going to follow that precedent.

chrislovecnm commented on incubator/traefik/templates/deployment.yaml:
```diff
@@ -0,0 +1,58 @@
+      labels:
+        k8s-app: traefik-ingress-controller
+        version: {{ .Chart.Version }}
+    spec:
```
ready and lively probes please, as well as resource limits ;)

krancour replied:
Yep. Good call.

chrislovecnm commented on incubator/traefik/templates/deployment.yaml:
```diff
@@ -0,0 +1,58 @@
+      {{- if .Values.acme.enabled }}
+      - name: acme
+      {{- if .Values.acme.persistence.enabled }}
+        persistentVolumeClaim:
```
It there value to have storage classes?

chrislovecnm replied:
It == Is

krancour replied:
I'm not sure what you're asking here. Can you clarify please?

This currently follows the exact pattern used by the stable MySQL chart for persistent storage.

viglesiasce replied:
The storage class is defined in the PVC not the deployment. Nothing to see here.

chrislovecnm reviewed (changes requested):
Thanks for the PR, looks awesome. Have some recommended tweaks

viglesiasce commented on incubator/traefik/values.yaml:
```diff
@@ -0,0 +1,20 @@
+ssl:
+  enabled: false
+  enforced: false
+  defaultCert: LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSUVtekNDQTRPZ0F3SUJBZ0lKQUpBR1FsTW1DMGt5TUEwR0NTcUdTSWIzRFFFQkJRVUFNSUdQTVFzd0NRWUQKVlFRR0V3SlZVekVSTUE4R0ExVUVDQk1JUTI5c2IzSmhaRzh4RURBT0JnTlZCQWNUQjBKdmRXeGtaWEl4RkRBUwpCZ05WQkFvVEMwVjRZVzF3YkdWRGIzSndNUXN3Q1FZRFZRUUxFd0pKVkRFV01CUUdBMVVFQXhRTktpNWxlR0Z0CmNHeGxMbU52YlRFZ01CNEdDU3FHU0liM0RRRUpBUllSWVdSdGFXNUFaWGhoYlhCc1pTNWpiMjB3SGhjTk1UWXgK ...
```
Can this be a secret name instead? Then the docs could outline how to create that secret (keys/values) such that it can be used here?

krancour replied:
If I can make an argument for keeping it as it is...

The base64 encoded cert and key you put here _becomes_ a secret. This means that you provide customizations to this either through `--set` on the command line or in your own `values.yaml`. i.e. You can customize these through the usual Helm workflow for customizing a chart.

Compare that to some other approach that requires a careful reading of the documentation and an "out of band" effort to create the secrets... I don't believe that compares favorably to what's in place now.

viglesiasce replied:
Fair. I'm fine as is. If users want something different, I'm sure they will chime in down the road.

Could I ask you to add instructions on getting those values? Also what do the default certs point to?

krancour replied:
The readme has a table describing the configuration options. It's clear that the cert and key should be base64 encoded. I can add more specific instructions if you like. lmk

The cert that's included is a self-signed cert for *.example.com-- so it's basically garbage. The benefit, however, is that it's enough to get people going.

I will update the table to reflect the details of this cert.

viglesiasce commented:
Awesome stuff @krancour!

viglesiasce commented:
@k8s-bot ok to test

ekozan commented:
@krancour  Can you integrate HostPort option or I'll make a PR after merge

ekozan commented on incubator/traefik/README.md:
````diff
@@ -0,0 +1,86 @@
+Specify each parameter using the `--set key=value[,key=value]` argument to `helm install`. For example:
+
+```bash
+$ helm install --name my-release --namespace kube-system --set dashboard.enabled=true,dashboard.domain=traefik.example.com stable/traefik
````
Actually it's incubator/traefik not stable/traefik

chrislovecnm replied:
If he adds a NOTES.txt it can be merged into stable directly

krancour commented:
@ekozan, did you mean `NodePort`? You want that as an alternative to `type: LoadBalancer`?

That seems a reasonable option (but not a default). It would certainly be a benefit to anyone who wants to run this anywhere where the "cloud provider" no-ops provisioning of an external load balancer-- e.g. minikube or bare metal.

I'd be happy to add that.

vdice commented on incubator/traefik/README.md:
````diff
@@ -0,0 +1,86 @@
+```
+
+After installing the chart, create a DNS CNAME record for applicable domains to direct inbound
+traffic the load balancer. You can you the command below to find the load balancer's DNS name:
````
a couple minor typos here; perhaps the following were intended: `... direct inbound traffic to the load balancer` and `You can use the command ...`

krancour replied:
Nice eye, @vdice. Will fix.

viglesiasce commented:
Tested manually and all things seem to check out. Thanks again @krancour! Let me know when you've patched the README fixes. Also if you'd like to get this into stable, adding a NOTES.txt pointing people in the right direction for how to configure and use traefik would do the trick.

krancour commented:
@viglesiasce great. I will try and amend with all suggested fixes as soon as time permits-- a few hours off, for sure.

ekozan commented:
Yeap it's an alternative for barebone cluster ;)

We need 
Add option for switch the default type of service: LoadBalancer / HostPort / NodePort  
Add option for switch to Deamonset is needed too :p i'll handle this if you dont want to add this one

ultimateboy commented:
Just thought I'd report my successful testing of this chart, and Traefik's ability to issue certs from Let's Encrypt.

```
helm install incubator/traefik --name my-traefik --namespace kube-system --set dashboard.enabled=true,dashboard.domain=traefik.domain.you.own.com,ssl.enabled=true,acme.enabled=true,acme.email=admin@domain.you.own.com
```

```
kubectl run nginx --image=nginx
kubecl expose deploy nginx --port=80 --target-port=80
```

```
echo "apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: nginx
  namespace: default
spec:
  rules:
  - host: some.domain.you.own.com
    http:
      paths:
      - backend:
          serviceName: nginx
          servicePort: 80" > ingress.yaml

kubectl create -f ingress.yaml
```

```
echo | openssl s_client -servername some.domain.you.own.com -connect some.domain.you.own.com:443 2>/dev/null | openssl x509 -noout -issuer
issuer= /C=US/O=Let's Encrypt/CN=Let's Encrypt Authority X3
```

krancour commented:
It'll be tomorrow before I get to this.

krancour committed:
Move traefik chart to stable

krancour committed:
Remove the license from the traefik chart

No other chart in this repo includes a license,
so I am following that precedent.

krancour committed:
Add resource limits for traefik

krancour committed:
Upgrade traefik to 1.1.0-rc3

krancour committed:
Add liveness and readiness probes for traefik

krancour committed:
Make traefik service type configurable

krancour committed:
Add missing documentation for imageTag option in Traefik chart

krancour committed:
Add support for Let's Encrypt's staging env

krancour commented:
@viglesiasce I've addressed comments apart from one. You indicated I should add a `NOTES.txt` "pointing people in the right direction for how to configure and use traefik."

I'm unclear on what that should contain or what its purpose would be.

`charts.yaml` already points users in the direction of the Traefik documentation. That aside, detailed knowledge of configuring Traefik isn't required to use this chart, since all customizations are carried out not by editing Traefik's config directly, but by supplying a `values.yaml` instead.

krancour commented:
@ultimateboy offline you asked me to support LE staging to avoid rate limits when testing / experimenting. I've added that now and made it the default as well (when ACME is enabled). Check it out and lmk what you think.

krancour committed:
Make Traefik resource limits configurable

krancour committed:
Fix typos in Traefik chart documentation

ultimateboy commented:
@krancour Awesome! It's probably worth explicitly calling that configuration out in the README as people might be confused why they aren't getting real valid certs. But this looks great!

krancour committed:
Add Traefik support for gzip compression

krancour committed:
Taming the metadata

krancour commented:
@viglesiasce nevermind re: my previous question about notes.txt... I realize now that this is an optional, but conventional component of charts. I've looked at some existing charts to get a sense of what sort of info is usually included. Pending the addition of those notes, I think this PR is into a reviewable state.

chrislovecnm commented:
Before we merge you mind squashing your commits into units of work? Thanks

krancour commented:
@chrislovecnm I was going to squash everything into one commit when done. Is that acceptable?

chrislovecnm commented:
@krancour your call mate!

viglesiasce commented:
No need to squash on your end as Github now allows Squash and merge when accepting the pull request.

krancour commented:
Ya... knew that, but not all projects necessarily have that enabled. If it's enabled here, awesome.

lachie83 commented:
I've reached out to @krancour and have agreed to work on getting a NOTES.txt into this chart.

lachie83 commented:
Seeing the following in the log file

```
time="2016-11-02T05:00:14Z" level=error msg="Kubernetes connection error failed to decode watch event: GET \"https://10.3.0.1:443/apis/extensions/v1beta1/ingresses\" : unexpected EOF, retrying in 682.444463ms" 
```

Looks like it's an issue with traefik -- https://github.com/containous/traefik/issues/732

lachie83 commented:
Tried putting an ingress in-front of Jenkins and it doesn't look like it handles more complex backends. Found this open issue -- https://github.com/containous/traefik/issues/30

Commit:
add NOTES.txt

lachie83 commented:
Added NOTES.txt. I believe this is the only outstanding item on this PR. Please review

krancour committed:
Update NOTES.txt

This commit includes formatting fixes and updates instructions to work
nicely across both AWS and GKE.

lachie83 commented:
@krancour Thanks for updating NOTES.txt. Looks good!

lachie83 commented:
@viglesiasce @prydonius @krancour I would like to propose that we reopen the discussion on moving this to the incubator. Based on my testing last night I found several things that didn't work and associated open issues on the Traefik repo. By putting this chart into stable may give the impression that it's ready for primetime. 

To sum up my thoughts here. The chart is stable, the application is still maturing.

krancour committed:
Final bit of polish on the README.md

lachie83 commented on stable/traefik/README.md:
```diff
@@ -8,7 +8,7 @@ microservices with ease.
 This chart bootstraps Traefik as a Kubernetes ingress controller with optional support for SSL and
 Let's Encrypt.
```
Can you a disclaimer about the state of Traefik?

krancour replied:
Yep. Meant to, as we'd discussed. Forgot. Coming atcha.

lachie83 commented on stable/traefik/README.md:
```diff
@@ -8,7 +8,7 @@ microservices with ease.
 This chart bootstraps Traefik as a Kubernetes ingress controller with optional support for SSL and
```
Looks great! Thanks for the updates!

lachie83 reviewed (changes requested):
Can we add a disclaimer on the current state of Traefik and it's prod readiness?

lachie83 reviewed (commented):
NOTES.txt looks much better! Thanks

prydonius commented:
@lachie83 @viglesiasce @krancour 

> By putting this chart into stable may give the impression that it's ready for primetime.

Agreed, if the project itself is not production-ready then the package should be in incubator.

krancour committed:
Add disclaimer about the state of Traefik; ymmv

lachie83 commented on stable/traefik/README.md:
```diff
@@ -3,6 +3,9 @@
 [Traefik](http://traefik.io/) is a modern HTTP reverse proxy and load balancer made to deploy
 microservices with ease.
 
+__DISCLAIMER:__ While this chart has been well-tested, testers have encountered occasional issues
```
Thanks @krancour

krancour commented:
To clarify, I believe Traefik regards itself as production ready and people certainly _do_ use Traefik in production. Any assertion about production readiness here is purely based on @lachie83's and my own hesitation to endorse this as such.

krancour commented:
So I'm going to move this to incubator. Standby.

krancour commented:
Moved

lachie83 commented:
> To clarify, I believe Traefik regards itself as production ready and people certainly do use Traefik in production. Any assertion about production readiness here is purely based on @lachie83's and my own hesitation to endorse this as such.

Agreed 👍

viglesiasce commented:
I dont know about that folks. Where is this production ready requirement coming from for stable? Its not listed anywhere and the other requirements are quite explicit. I don't want us to cross this line.

krancour commented:
So if the chart's own author(s) would hesitate to use it in production (depending on the use case), this can still qualify as "stable?"

krancour commented:
Maybe some clarification on what, precisely, is meant by "stable" is in order. I will acknowledge I made a pretty big (and perhaps incorrect) assumption that stability was equated with production readiness.

lachie83 commented:
> I dont know about that folks. Where is this production ready requirement coming from for stable? Its not listed anywhere and the other requirements are quite explicit. I don't want us to cross this line.

Understood and I agree that it's not our position to decide whether something is production ready. I don't want to cross that line. Stable does not mean Prod Ready

How about we approach this from a difference angle - putting this in the incubator is a great way for it to gather some miles

viglesiasce commented:
Stable charts meet [these requirements](https://github.com/kubernetes/charts/blob/master/CONTRIBUTING.md#how-to-contribute-a-chart).

If we cross the line of requiring things to be production ready then we need to define production ready in a way that is obvious otherwise we are going to end up with a lot of bikeshedding. For most packages (rpm and deb) the defaults are sane but not production ready.

krancour commented:
> How about we approach this from a difference angle - putting this in the incubator is a great way for it to gather some miles

@viglesiasce is there any validity to that notion? That things are "promoted" so stable after proving themselves in the incubator?

viglesiasce commented:
The only thing we say in the [README.md](https://github.com/kubernetes/charts#repository-structure) is that stable charts meet the criteria in the [technical requirements](https://github.com/kubernetes/charts/blob/master/CONTRIBUTING.md#how-to-contribute-a-chart). If we want to change that policy then we need to update the documentation via a PR where we can hash out the why and how.

krancour commented:
Ok. I'm going to strike the last commit (the move to incubator) from this PR.

lachie83 commented:
Thanks for the great discussion and clarification. 

Given that this chart meets the requirements. My vote is on putting this in stable.

krancour commented:
All set. Back in stable.

prydonius commented:
@viglesiasce you're right, for some reason I thought some level of 'production-readiness' was a requirement. I would argue for promoting the stable repo as a source of 'production-ready' charts, but it's a separate discussion to define what we could mean by that and whether it makes sense to enforce it.

lachie83 commented:
LGTM

viglesiasce commented:
LGTM

prydonius commented:
@krancour lgtm, all good to merge?

krancour commented:
Yes

lachie83 commented:
Thanks all! Excited to have this chart!
